Preprocessing/Base/converter.py

```diff
@@ -0,0 +1,311 @@
+    # channel_names = signals.ch_names
+    #
+    # placeholder_data = []
+    # for eeg_data in eeg_data_list:
+    #     placeholder_data.append(resample_list(eeg_data))
+    #
+    # channel_types = ['eeg'] * len(channel_names)
+    #
+    # info = mne.create_info(ch_types=channel_types, sfreq=new_sample_rate, ch_names=channel_names)
+    # raw_data = mne.io.RawArray(placeholder_data, info)
+    # return raw_data
+
+
+def scale_values(signals: mne.io.RawArray | RawEDF, min_val: float = -1, max_val: float = 1) -> mne.io.RawArray:
+    """
+    Scales all values to range [a,b]
+
+    :param signals: Signals in RawEdf or RawArray
+    :param min_val: Lower bound of range
+    :param max_val: Upper bound of range
+    :return:
+    """
+    data = signals.get_data()
+    data_normalized = (data - np.min(data)) / (np.max(data) - np.min(data))
+    data_normalized = data_normalized * (max_val - min_val) + min_val
+
+    info = mne.create_info(ch_types=signals.get_channel_types(), sfreq=signals.info['sfreq'], ch_names=signals.ch_names)
+    info.set_montage('standard_1020')
+    raw_data = mne.io.RawArray(data_normalized, info)
+    return raw_data
+
+
+def split_into_time_windows(signals: mne.io.RawArray, sample_frequency: int, secs: float = 3) -> Generator[
+    dict[str,np.ndarray[float]], None, None]:
+    """
+    Returns some time window from signal. It's generator for simplicity’s sake.
+
+    :param signals: RawArray from mne with all signals
+    :param sample_frequency: Frequency at which data is sampled.
+    :param secs: By what time divide data
+    :return: Signal from time chunk
+    """
+
+    epochs = mne.make_fixed_length_epochs(signals, duration=secs, preload=True)
+
+    for chunk in epochs:
+        # zips chunks into dict
+        yield dict(zip(epochs.ch_names,chunk))
+
+    # data = signals.get_data()
+    # time = len(data[0]) / sample_frequency
+    #
+    # chunks = int(time // 3)
+    #
+    # for i in range(chunks - 1):
+    #     # [chunk_start : chunk_end]
+    #     chunk_start = int((secs * sample_frequency) * i)
+    #     chunk_end = int((secs * sample_frequency) * (i + 1))
+    #     mapped_data = list(map(lambda x: x[chunk_start:chunk_end], data))
+    #     info = mne.create_info(
+    #         ch_types=signals.get_channel_types(),
+    #         sfreq=signals.info['sfreq'],
+    #         ch_names=signals.ch_names)
+    #     raw_data = mne.io.RawArray(mapped_data, info)
+    #
+    #     yield raw_data
+    #
+    # mapped_data = list(map(lambda x: x[int((secs * sample_frequency) * (chunks - 1)):], data))
+    # info = mne.create_info(
+    #     ch_types=signals.get_channel_types(),
+    #     sfreq=signals.info['sfreq'],
+    #     ch_names=signals.ch_names
+    # )
+    # info.set_montage('standard_1020')
+    # raw_data = mne.io.RawArray(mapped_data, info)
+    #
+    # yield raw_data
+
+
+def process_folder(folder_name: str, mode: Literal['edf', 'csv', 'eea'], output_folder: str = None) -> None:
```

Comment:
Jeśli chodzi o dumpowanie plików, ma ono na celu poradzenie sobie z problemem wolnego odczytu danych z dysku twardego. Zamiast odczytywać tysiące (w naszym przypadku tylko dziesiątki) plików -> otwieramy tylko jeden plik, wykonujey jeden duży odczyt zdumpowanych danych. Czy tam max 3 czy 6 odczytów w naszym przypadku (3 datasety * 2 klasy).
Zamiast zapisywać każdą 3 sekundową część nagrania w osobnym pliku `.pk` -> zapiszmy wszystkie pliki z jednego datasetu z jednej klasy (np. chory) do jednego pliku `.pk`: `eeg_schizophrenia.pk`, choć uwzględniając dataset to będzie np.: `eeg_csv_schizophrenia.pk`.

Zobacz na plik:
https://github.com/EEG-PK/schizophrenia-classification/blob/development/model/mockup_data.py
W funkcji `mockup` wszystkie sygnały osób zdrowych są zapisywane do jednego pliku `.pk`, i do drugiego pliku `.pk` wszystkie sygnały osób chorych. Struktura pliku `.pk` to tablica słowników, gdzie w każdym słowniku znajdują się dane jednego sygnału eeg (jednej osoby), pod kluczem 'eeg'.
Funkcja `read_eeg_data` zapisuje obecnie zwykłą dwuwymiarową tablicę (bez nazw kanałów) - trzeba dodać nazwy każdego kanału, coś na tej zasadzie:
```
    # Create a dictionary with channel names as keys
    channels_data = {}
    for i, channel in enumerate(common_channels):
        start_index = i * num_samples_per_channel
        end_index = (i + 1) * num_samples_per_channel
        channel_data = df.iloc[start_index:end_index, 0].values
        channels_data[channel] = channel_data

    return channels_data
```
Czyli stworzyć kolejny słownik z nazwami kanałów, który zamieścimy w kluczu `eeg`.

Nie musisz używać tego kodu powyżej, lepiej edytować już gotowy kod w tym pliku i skorzystać z tego, że mamy nazwy kanałów w obiekcie mne - i je dodać do słownika. I potem tablicę takich słowników z danymi sygnałów osób np. chorych dumpujemy do pliku `.pk`.
Najlepiej dumpować wszystko jeszcze przed podziałem plików na segmenty (to już będzie edytowane po stronie pipeline'u modelu, gdzie będę mogła to potestować bez potrzeby zapisu danych ponownie na dysk). Czyli podział na segmenty można na ten moment usunąć/zakomentować.

Czyli na koniec mamy pliki takie jak:
`eeg_csv_schizophrenia.pk`
`eeg_csv_health.pk`
`eeg_edf_schizophrenia.pk`
`eeg_edf_health.pk`
`eeg_eea_schizophrenia.pk`
`eeg_eea_health.pk`

Reply:
Zmieniłem na to aby zapisywało się do jednego pliku z odpowiednimi nazwami plików. Do tego dodałem progress bar (ze stacka) aby było widoczne jaki jest progress plików. Pliki eea robią się w niecałe 5 min wszystkie, csv jest ciężej jeden plik robi się około 6-7 min co daje nam około 8 h preprocessingu dla całego datasetu, a edf nie sprawdzałem na moment pisania.